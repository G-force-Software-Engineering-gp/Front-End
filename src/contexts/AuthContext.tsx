import { createContext, useState, useEffect } from "react";
import jwt_decode from "jwt-decode";
import { useNavigate } from "react-router-dom";
import axios from "axios";
// import swal from "sweetalert";
// import axiosInstance from "../utils/axiosInstance";


export type AuthContextType = {
  user: any; // Replace with your user type
  setUser: React.Dispatch<React.SetStateAction<any>>;
  authTokens: any; // Replace with your auth tokens type
  setAuthTokens: React.Dispatch<React.SetStateAction<any>>;
  loginUser: (values: any) => Promise<void>;
  registerUser: (values: any) => Promise<void>;
  logoutUser: () => void;
};

const AuthContext = createContext<AuthContextType | null>(null);

export default AuthContext;
type AuthProviderProps = {
    children: React.ReactNode;
  };
  
  export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const [authTokens, setAuthTokens] = useState<any>(
    localStorage.getItem("authTokens")
      ? JSON.parse(localStorage.getItem("authTokens")!)
      : null
  );
  const [user, setUser] = useState<any>(
    localStorage.getItem("authTokens")
      ? jwt_decode(localStorage.getItem("authTokens")!)
      : null
  );
  const [loading, setloading] = useState(true);

  const Navigate = useNavigate();

  const loginUser = async (values: any) => {
    const response = await fetch(
        "https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            email: values.email,
            password: values.password,
          }),
        }
      );
      const data = await response.json();
      console.log(data)
      if (response.status === 200) {
        setAuthTokens(data);
        setUser(jwt_decode(data.access));
        localStorage.setItem("authTokens", JSON.stringify(data));
        Navigate("/");
      } 
      else{
        console.log(data.error);
      }
  };
  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        //     {
        //     email: values.email,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((response) => {
        //     const data = response.data;
        //     console.log(data);
        
        //     if (response.status === 200) {
        //         setAuthTokens(data);
        //         setUser(jwt_decode(data.access));
        //         localStorage.setItem('authTokens', JSON.stringify(data));
        //         Navigate('/');
        //     } else {
        //         console.log(data.error);
        //     }
        //     })
        //     .catch((error) => {
        //     console.error('API request failed:', error);
        //     });
        // }
  const registerUser = async (values: any) => {
      const reg_response = await fetch(
      "https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
      {
          method: "POST",
          headers: {
          "Content-Type": "application/json",
          },
          body: JSON.stringify({
          first_name: values.firstName,
          last_name: values.lastName,
          email: values.email,
          username: values.username,
          password: values.password,
          }),
      });
      console.log(reg_response);
      const reg_data = await reg_response.json();
      if (reg_response.status === 201) {
        Navigate("/login");
      } else {
          console.log(reg_data.error);  
      }
  };

  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
        //     {
        //     first_name: values.firstName,
        //     last_name: values.lastName,
        //     email: values.email,
        //     username: values.username,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((reg_data) => {
        //         const data = reg_data.data;
        //         console.log(data);
        //         if (reg_data.status === 201) {
        //           Navigate('/login');
        //         } else {
        //           console.log(data.error);
        //         }
        //       })
        //       .catch((error) => {
        //         console.error('API request failed:', error);
        //       });
        // }

  const logoutUser = () => {
    setAuthTokens(null);
    setUser(null);
    localStorage.removeItem("authTokens");
    Navigate("/login");
  };

  const contextData: AuthContextType = {
    user,
    setUser,
    authTokens,
    setAuthTokens,
    loginUser,
    registerUser,
    logoutUser,
  };

  useEffect(() => {
    if (authTokens) {
        setUser(jwt_decode(authTokens.access));
      }
      setloading(false);
  }, [authTokens, loading]);

  return <AuthContext.Provider value={contextData}>{children}</AuthContext.Provider>;
};
