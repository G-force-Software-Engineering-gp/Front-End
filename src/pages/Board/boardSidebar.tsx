import { Avatar, AvatarFallback, AvatarImage } from '@/components/ui/avatar';
import { Button } from '@/components/ui/button';
import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem, CommandList } from '@/components/ui/command';
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog';
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';
import AuthContext from '@/contexts/AuthContext';
import { cn } from '@/lib/utils';
import { useQueryClient } from '@tanstack/react-query';
import axios from 'axios';
import _ from 'lodash';
import { CalendarDays, Check, ChevronRight, Plus, Settings, Trash2, Trello, Users2 } from 'lucide-react';
import { useContext, useEffect, useRef, useState } from 'react';
import { useParams } from 'react-router-dom';
import { BaseURL } from '../baseURL';
import { useBoard } from './hooks/useBoard';

interface User {
  id: number;
  username: string;
  first_name: string;
  last_name: string;
  email: string;
  member: Member[];
}

interface Member {
  profimage: string;
}

export function BoardSidebar({ className }: any) {
  const { boardId } = useParams();
  const [addMemberButtonLoading, setAddMemberButtonLoading] = useState(false);
  const queryClient = useQueryClient();
  function inviteMembers() {
    const fetches = selectedUsers
      .map((item) => {
        return { member: item.id, board: boardId };
      })
      .map((item) => {
        return fetch(BaseURL + `tascrum/invite/`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `JWT ` + authTokens.access,
          },
          body: JSON.stringify(item),
        });
      });
    Promise.allSettled(fetches)
      .then((results) =>
        results.forEach((item) => {
          setOpen(false);
          setSelectedUsers([]);
        })
      )
      .then((res) => {
        queryClient.invalidateQueries({ queryKey: ['members'] });
      });
  }
  let authTokens = useContext(AuthContext)?.authTokens;
  const [query, setQuery] = useState('');
  const deferredQuery = query;

  const [users, setUsers] = useState<User[] | []>([]);

  const [selectedUsers, setSelectedUsers] = useState<User[]>([]);

  const allUsers = useRef<User[] | []>([]);

  useEffect(() => {
    if (deferredQuery !== '') {
      axios
        .get(BaseURL + `tascrum/user-search/?board=${boardId}&search=${deferredQuery}`, {
          headers: {
            Authorization: `JWT ${authTokens.access}`,
          },
        })
        .then((res) => {
          const fetchedUsers: User[] = res.data;
          setUsers(fetchedUsers);
          allUsers.current = _.unionBy(allUsers.current, fetchedUsers, 'id');
        });
    } else {
      setUsers(selectedUsers);
    }
  }, [deferredQuery]);

  useEffect(() => {
    if (deferredQuery === '' && selectedUsers.length === 0) {
      setUsers([]);
    }
  }, [deferredQuery, selectedUsers]);

  const [open, setOpen] = useState(false);
  return (
    <div className={cn('pb-12', className)}>
      <div className="space-y-4 py-4 ">
        <div className="px-3 py-2">
          <div className="space-y-1">
            <Button data-testid="Boards" variant="ghost" className="w-full justify-start">
              <Trello className="mr-2 h-4 w-4" />
              Boards
            </Button>
            <Button data-testid="Members" variant="ghost" className="group w-full justify-between pr-2">
              <div className="flex justify-start">
                <Users2 className="mr-2 h-4 w-4" />
                Members
              </div>
              <Dialog open={open} onOpenChange={setOpen}>
                <DialogTrigger asChild>
                  <Button variant="outline" className="hidden h-7  w-7 p-0 group-hover:flex">
                    <Plus className="h-5 w-5" />
                  </Button>
                </DialogTrigger>
                <DialogContent className="gap-0 p-0 outline-none">
                  <DialogHeader className="px-4 pb-4 pt-5">
                    <DialogTitle>Invite to Workspace</DialogTitle>
                    <DialogDescription>Invite a user to this Board.</DialogDescription>
                  </DialogHeader>
                  <Command shouldFilter={false} className="overflow-hidden rounded-t-none border-t">
                    <CommandInput
                      placeholder="Search user..."
                      value={query}
                      onValueChange={(str) => {
                        setQuery(str);
                      }}
                    />
                    <CommandList>
                      <CommandEmpty>No users found.</CommandEmpty>
                      {users.length !== 0 && (
                        <CommandGroup className="p-2">
                          {users.length !== 0 &&
                            users.map((user) => (
                              <CommandItem
                                key={user.email}
                                className="flex items-center px-2"
                                onSelect={() => {
                                  if (selectedUsers.filter((e: User) => e.email === user.email).length > 0) {
                                    return setSelectedUsers(
                                      selectedUsers.filter((selectedUser) => selectedUser.email !== user.email)
                                    );
                                  }

                                  const moteghayer = _.intersectionBy(allUsers.current, [...selectedUsers, user], 'id');
                                  return setSelectedUsers(moteghayer);
                                }}
                              >
                                <Avatar>
                                  <AvatarImage src={user.member[0].profimage} alt="Image" />
                                  <AvatarFallback>
                                    {user.first_name[0]}
                                    {user.last_name[0]}
                                  </AvatarFallback>
                                </Avatar>
                                <div className="ml-2">
                                  <p className="text-sm font-medium leading-none">
                                    {user.first_name} {user.last_name}
                                  </p>
                                  <p className="text-sm text-muted-foreground">{user.email}</p>
                                </div>
                                {selectedUsers.filter((item) => item.email === user.email).length > 0 ? (
                                  <Check className="ml-auto flex h-5 w-5 text-primary" />
                                ) : null}
                              </CommandItem>
                            ))}
                        </CommandGroup>
                      )}
                    </CommandList>
                  </Command>
                  <DialogFooter className="flex items-center border-t p-4 sm:justify-between">
                    {selectedUsers.length > 0 ? (
                      <div className="flex -space-x-2 overflow-hidden">
                        {selectedUsers.map((user) => (
                          <Avatar key={user.email} className="inline-block border-2 border-background">
                            <AvatarImage src={user.member[0].profimage} />
                            <AvatarFallback>
                              {user.first_name[0]}
                              {user.last_name[0]}
                            </AvatarFallback>
                          </Avatar>
                        ))}
                      </div>
                    ) : (
                      <p className="text-sm text-muted-foreground">Select users to add to this Board.</p>
                    )}
                    <Button
                      disabled={selectedUsers.length < 1 || addMemberButtonLoading}
                      onClick={() => {
                        inviteMembers();
                      }}
                    >
                      {addMemberButtonLoading && <p>Loading...</p>}
                      {!addMemberButtonLoading && <p>Add Members</p>}
                    </Button>
                  </DialogFooter>
                </DialogContent>
              </Dialog>
            </Button>
            <DropdownMenu>
              <DropdownMenuTrigger asChild>
                <Button variant="ghost" className="h-fit w-full justify-between pr-3">
                  <div className="flex items-center justify-start">
                    <Settings className="mr-2 h-4 w-4" />
                    Workspace Settings
                  </div>
                  <ChevronRight className="h-5 w-5" />
                </Button>
              </DropdownMenuTrigger>
              <DropdownMenuContent className="w-56">
                <DropdownMenuGroup>
                  <DropdownMenuItem>
                    <span>Workspace settings</span>
                  </DropdownMenuItem>
                  <DropdownMenuItem>
                    <span>Upgrade Workspace</span>
                  </DropdownMenuItem>
                </DropdownMenuGroup>
              </DropdownMenuContent>
            </DropdownMenu>
          </div>
        </div>
        <div className="px-3 py-2">
          <h2 className="mb-2 px-4 text-lg font-semibold tracking-tight">Workspace Views</h2>
          <div className="space-y-1">
            <Button variant="ghost" className="group w-full justify-between pr-2">
              <div className="flex justify-start">
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  width="24"
                  height="24"
                  viewBox="0 0 24 24"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  className="lucide lucide-table-properties mr-2 h-4 w-4"
                >
                  <path d="M15 3v18" />
                  <rect width="18" height="18" x="3" y="3" rx="2" />
                  <path d="M21 9H3" />
                  <path d="M21 15H3" />
                </svg>
                <span className="italic">Table</span>
              </div>
              <Button variant="secondary" className="hidden h-7 w-7 p-0 group-hover:flex">
                <Trash2 className="hidden h-4 w-4 group-hover:block" />
              </Button>
            </Button>
            <Button variant="ghost" className="group w-full justify-between pr-2">
              <div className="flex justify-start">
                <CalendarDays className="mr-2 h-4 w-4" />
                <span className="italic">Calendar</span>
              </div>
              <Button variant="secondary" className="hidden h-7  w-7 p-0 group-hover:flex">
                <Trash2 className="hidden h-4 w-4 group-hover:block" />
              </Button>
            </Button>
          </div>
        </div>
        <div className="px-3 py-2">
          <h2 className="mb-2 px-4 text-lg font-semibold tracking-tight">Your Boards</h2>
          <div className="space-y-1">
            <Button variant="secondary" className="w-full justify-start">
              <svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 24 24"
                fill="none"
                stroke="currentColor"
                strokeWidth="2"
                strokeLinecap="round"
                strokeLinejoin="round"
                className="mr-2 h-4 w-4"
              >
                <path d="M21 15V6" />
                <path d="M18.5 18a2.5 2.5 0 1 0 0-5 2.5 2.5 0 0 0 0 5Z" />
                <path d="M12 12H3" />
                <path d="M16 6H3" />
                <path d="M12 18H3" />
              </svg>
              Table
            </Button>
          </div>
        </div>
      </div>
    </div>
  );
}
