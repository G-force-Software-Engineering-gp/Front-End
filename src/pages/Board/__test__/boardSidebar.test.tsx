import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import { BoardSidebar } from '../boardSidebar';
import '@testing-library/jest-dom/extend-expect'; // for expect assertions

import { AuthProvider } from '@/contexts/AuthContext';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter } from 'react-router-dom';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});
describe('BoardSidebar component', () => {
  test('renders BoardSidebar component with correct texts and placeholders', () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <AuthProvider>
            <BoardSidebar />
          </AuthProvider>
        </QueryClientProvider>
      </MemoryRouter>
    );

    expect(screen.getByText('Boards')).toBeInTheDocument();
    expect(screen.getByText('Members')).toBeInTheDocument();
    expect(screen.getByText('Workspace Views')).toBeInTheDocument();
    expect(screen.getByText('Calendar')).toBeInTheDocument();
    expect(screen.getByText('Your Boards')).toBeInTheDocument();
  });

  test('renders correct roles for accessibility', () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <AuthProvider>
            <BoardSidebar />
          </AuthProvider>
        </QueryClientProvider>
      </MemoryRouter>
    );

    expect(screen.getByTestId(/Boards/i)).toBeInTheDocument();
    expect(screen.getByTestId(/Members/i)).toBeInTheDocument();
  });
});

// Mock the dependencies
jest.mock('axios');
jest.mock('lodash');

describe('BoardSidebar Component', () => {
  // Rendering Test
  test('renders BoardSidebar component', () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <AuthProvider>
            <BoardSidebar />
          </AuthProvider>
        </QueryClientProvider>
      </MemoryRouter>
    );
    // You can add more specific assertions based on your UI components
    expect(screen.getByTestId('Boards')).toBeInTheDocument();
    expect(screen.getByTestId('Members')).toBeInTheDocument();
  });

  // Interaction Test
  test('clicking on the invite button opens the dialog', async () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <AuthProvider>
            <BoardSidebar />
          </AuthProvider>
        </QueryClientProvider>
      </MemoryRouter>
    );

    // Ensure the dialog is initially closed
    expect(screen.queryByText('Invite to Workspace')).not.toBeInTheDocument();

    // Trigger the dialog open
    fireEvent.click(screen.getByTestId('Members'));
  });

  // Snapshot Test
  // test('matches snapshot', () => {
  //   const { asFragment } = render(<BoardSidebar />);
  //   expect(asFragment()).toMatchSnapshot();
  // });
});
