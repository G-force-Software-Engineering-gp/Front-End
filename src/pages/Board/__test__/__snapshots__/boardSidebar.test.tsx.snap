// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardSidebar Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="pb-12"
  >
    <div
      class="space-y-4 py-4 "
    >
      <div
        class="px-3 py-2"
      >
        <div
          class="space-y-1"
        >
          <button
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 w-full justify-start"
            data-testid="Boards"
          >
            <svg
              class="mr-2 h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="3"
              />
              <rect
                height="9"
                width="3"
                x="7"
                y="7"
              />
              <rect
                height="5"
                width="3"
                x="14"
                y="7"
              />
            </svg>
            Boards
          </button>
          <button
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 group pr-2 w-full justify-between"
            data-testid="Members"
          >
            <div
              class="flex justify-start"
            >
              <svg
                class="mr-2 h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M14 19a6 6 0 0 0-12 0"
                />
                <circle
                  cx="8"
                  cy="9"
                  r="4"
                />
                <path
                  d="M22 19a6 6 0 0 0-6-6 4 4 0 1 0 0-8"
                />
              </svg>
              Members
            </div>
          </button>
          <button
            aria-controls="radix-:rk:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input hover:bg-accent hover:text-accent-foreground hidden group-hover:flex h-7 w-7 p-0"
            data-state="closed"
            type="button"
          >
            <svg
              class="w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="5"
                y2="19"
              />
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground px-4 py-2 w-full h-fit justify-between pr-3"
            data-state="closed"
            id="radix-:rn:"
            type="button"
          >
            <div
              class="flex justify-start items-center"
            >
              <svg
                class="mr-2 h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12.22 2h-.44a2 2 0 0 0-2 2v.18a2 2 0 0 1-1 1.73l-.43.25a2 2 0 0 1-2 0l-.15-.08a2 2 0 0 0-2.73.73l-.22.38a2 2 0 0 0 .73 2.73l.15.1a2 2 0 0 1 1 1.72v.51a2 2 0 0 1-1 1.74l-.15.09a2 2 0 0 0-.73 2.73l.22.38a2 2 0 0 0 2.73.73l.15-.08a2 2 0 0 1 2 0l.43.25a2 2 0 0 1 1 1.73V20a2 2 0 0 0 2 2h.44a2 2 0 0 0 2-2v-.18a2 2 0 0 1 1-1.73l.43-.25a2 2 0 0 1 2 0l.15.08a2 2 0 0 0 2.73-.73l.22-.39a2 2 0 0 0-.73-2.73l-.15-.08a2 2 0 0 1-1-1.74v-.5a2 2 0 0 1 1-1.74l.15-.09a2 2 0 0 0 .73-2.73l-.22-.38a2 2 0 0 0-2.73-.73l-.15.08a2 2 0 0 1-2 0l-.43-.25a2 2 0 0 1-1-1.73V4a2 2 0 0 0-2-2z"
                />
                <circle
                  cx="12"
                  cy="12"
                  r="3"
                />
              </svg>
              Workspace Settings
            </div>
            <svg
              class="w-5 h-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="9 18 15 12 9 6"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="px-3 py-2"
      >
        <h2
          class="mb-2 px-4 text-lg font-semibold tracking-tight"
        >
          Workspace Views
        </h2>
        <div
          class="space-y-1"
        >
          <button
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 group w-full justify-between pr-2"
          >
            <div
              class="flex justify-start"
            >
              <svg
                class="mr-2 h-4 w-4 lucide lucide-table-properties"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15 3v18"
                />
                <rect
                  height="18"
                  rx="2"
                  width="18"
                  x="3"
                  y="3"
                />
                <path
                  d="M21 9H3"
                />
                <path
                  d="M21 15H3"
                />
              </svg>
              <span
                class="italic"
              >
                Table
              </span>
            </div>
          </button>
          <button
            class="items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground hover:bg-secondary/80 hidden group-hover:flex h-7 w-7 p-0"
          >
            <svg
              class="h-4 w-4 hidden group-hover:block"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 6h18"
              />
              <path
                d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
              />
              <path
                d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 group w-full justify-between pr-2"
          >
            <div
              class="flex justify-start"
            >
              <svg
                class="mr-2 h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <line
                  x1="16"
                  x2="16"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="8"
                  x2="8"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="3"
                  x2="21"
                  y1="10"
                  y2="10"
                />
                <path
                  d="M8 14h.01"
                />
                <path
                  d="M12 14h.01"
                />
                <path
                  d="M16 14h.01"
                />
                <path
                  d="M8 18h.01"
                />
                <path
                  d="M12 18h.01"
                />
                <path
                  d="M16 18h.01"
                />
              </svg>
              <span
                class="italic"
              >
                Calendar
              </span>
            </div>
          </button>
          <button
            class="items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground hover:bg-secondary/80 hidden group-hover:flex h-7 w-7 p-0"
          >
            <svg
              class="h-4 w-4 hidden group-hover:block"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 6h18"
              />
              <path
                d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
              />
              <path
                d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="px-3 py-2"
      >
        <h2
          class="mb-2 px-4 text-lg font-semibold tracking-tight"
        >
          Your Boards
        </h2>
        <div
          class="space-y-1"
        >
          <button
            class="inline-flex items-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-4 py-2 w-full justify-start"
          >
            <svg
              class="mr-2 h-4 w-4"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 15V6"
              />
              <path
                d="M18.5 18a2.5 2.5 0 1 0 0-5 2.5 2.5 0 0 0 0 5Z"
              />
              <path
                d="M12 12H3"
              />
              <path
                d="M16 6H3"
              />
              <path
                d="M12 18H3"
              />
            </svg>
            Table
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
